v2/api/authorization/customizations/role_assignment_extensions.go

```diff
@@ -43,3 +56,126 @@ func (extension *RoleAssignmentExtension) Import(
 
 	return result, nil
 }
+
+var _ extensions.ARMResourceModifier = &RoleAssignmentExtension{}
+
+func (extension *RoleAssignmentExtension) ModifyARMResource(
+	ctx context.Context,
+	armClient *genericarmclient.GenericClient,
+	armObj genruntime.ARMResource,
+	obj genruntime.ARMMetaObject,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) (genruntime.ARMResource, error) {
+	ra, ok := obj.(*storage.RoleAssignment)
+	if !ok {
+		return nil, eris.Errorf(
+			"Cannot run RoleAssignmentExtension.ModifyARMResource() with unexpected resource type %T",
+			obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not been updated to match
+	var _ conversion.Hub = ra
+
+	// If the specified role definition uses a well known name, look it up
+	roleDefinitionName := ra.Spec.RoleDefinitionReference.WellKnownName
+	if roleDefinitionName != "" {
+		err := ensureBuiltInRoleDefinitionsLoaded(ctx, armClient)
+		if err != nil {
+			return nil, eris.Wrapf(err, "loading built in role definitions to resolve %q", roleDefinitionName)
+		}
+
+		roleDefinitionId, err := resolveBuiltInRoleDefinition(roleDefinitionName, armObj)
+		if err != nil {
+			return nil, eris.Wrapf(err, "resolving built in role definition %q", roleDefinitionName)
+		}
+
+		if roleDefinitionId != roleDefinitionName {
+			log.V(1).Info("Resolved built-in role", "roleName", roleDefinitionName, "roleId", roleDefinitionId)
+
+			spec := armObj.Spec()
+			err = reflecthelpers.SetProperty(spec, "Properties.RoleDefinitionId", &roleDefinitionId)
+			if err != nil {
+				return nil, eris.Wrapf(err, "error setting RoleDefinitionId to %s", roleDefinitionId)
+			}
+
+			return armObj, nil
+		}
+	}
+
+	return armObj, nil
+}
+
+var (
+	// builtInRoleDefinitions is a cache of all known built-in role definitions, keyed by
+	// lower case role name. At the time of writing there are ~700 such roles, so we
+	// pre-initialize the map to sufficient capacity to accommodate those and some modest growth.
+	builtInRoleDefinitions map[string]string = make(map[string]string, 800)
+
+	// builtInRoleDefinitionsLock protects access to builtInRoleDefinitions
+	builtInRoleDefinitionsLock sync.RWMutex
+)
+
+func ensureBuiltInRoleDefinitionsLoaded(
+	ctx context.Context,
+	armClient *genericarmclient.GenericClient,
+) error {
+	builtInRoleDefinitionsLock.Lock()
+	defer builtInRoleDefinitionsLock.Unlock()
+
+	// Short circuit if already loaded
```

Comment:
minor: you could double-check this (before/after lock). Advantage is that it avoids taking the lock at all after the first time which is nice.

It's a small optimization though so feel free to ignore if you don't think it's worth it. If you do do this, I'd leave a comment explaining the 2x check.